src/api/controllers/db-controller.ts

```diff
@@ -828,43 +939,153 @@ export async function getTxFromDataStore(
-      ).map((c, fnArgIndex) => {
-        const functionArgAbi = functionAbi.args[fnArgIndex++];
-        return {
-          hex: bufferToHexPrefixString(serializeCV(c)),
-          repr: cvToString(c),
-          name: functionArgAbi.name,
-          type: getTypeString(functionArgAbi.type),
-        };
-      });
-    }
   }
+  parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
+  if (dbTransaction.contract_call_function_args) {
+    parsedTx.contract_call.function_args = readClarityValueArray(
+      dbTransaction.contract_call_function_args
+    ).map((c, fnArgIndex) => {
+      const functionArgAbi = functionAbi.args[fnArgIndex++];
+      return {
+        hex: bufferToHexPrefixString(serializeCV(c)),
+        repr: cvToString(c),
+        name: functionArgAbi.name,
+        type: getTypeString(functionArgAbi.type),
+      };
+    });
+  }
+  return parsedTx;
+}
+
+export async function searchTxs(
+  db: DataStore,
+  args: GetTxsArgs | GetTxsWithEventsArgs
+): Promise<FoundOrNot<Transaction | MempoolTransaction | string>[]> {
+  const minedTxs = await getTxsFromDataStore(db, args);
+
+  // filtering out mined transactions in canonical chain
+  const minedAndInCanonicalChain = minedTxs.filter(tx => tx.canonical && tx.microblock_canonical);
+
+  // filtering out tx_ids that are not in canonical chain
+  const nonCanonical = minedTxs.filter(tx => !tx.canonical && !tx.microblock_canonical);
+
+  // filtering out tx_ids that were not mined / found
+  const notMinedTransactions: string[] = args.txIds.filter(
+    txId => !minedTxs.find(minedTx => txId === minedTx.tx_id)
+  );
+
+  // finding transactions that are not mined and are not in canonical in mempool.
+  const mempoolTxs = [...nonCanonical.map(tx => tx.tx_id), ...notMinedTransactions];
+  const mempoolTxsQuery = await getMempoolTxsFromDataStore(db, {
+    txIds: mempoolTxs,
+    includeUnanchored: args.includeUnanchored,
+  });
+
+  const foundOrNotMined = minedAndInCanonicalChain.map(minedTx => {
+    return { found: true, result: minedTx };
+  });
+  const foundOrNotMempool = mempoolTxsQuery.map((mtx: any) => {
+    return { found: true, result: mtx };
+  });
+  // such transactions that are not in mempoolTxQuery but do exist in notInCanonical
+  const foundOrNotNonCanonical = nonCanonical
+    .filter(tx => mempoolTxsQuery.findIndex((mtx: any) => mtx.tx_id === tx.tx_id) < 0)
+    .map(tx => {
+      return { found: true, result: tx };
+    });
+  // all transactions that were not found anywhere
+  const foundOrNotFoundTxs = [...foundOrNotMined, ...foundOrNotMempool, ...foundOrNotNonCanonical];
+  const foundOrNotNotFoundTxs = args.txIds
+    .filter(txId => foundOrNotFoundTxs.findIndex(ftx => ftx.result.tx_id === txId) < 0)
+    .map(txId => {
+      return { found: false, result: txId };
+    });
+
+  // sorting according to incoming tx_ids
+  const resp = [...foundOrNotFoundTxs, ...foundOrNotNotFoundTxs];
+  resp.sort((tx1, tx2) => {
```

Comment:
Is this sorting really necessary? Seems like a very expensive operation that doesn't provide too much value to the response. If your goal is to allow users to get transaction results by index, why not switch to an associated array response? It would be much easier to use and much less expensive to produce.

For example, instead of returning this for each transaction:
```
{
  "found": true,
  "result": {
    "tx_id": "0x1234"
  }
}
```
You can do this:
```
{
  "0x1234": {
    "found": true,
    "result": {
      "tx_id": "0x1234"
    }
  }
  (...other txs here)
}
```
This way, users can parse the object and quickly search each tx by tx_id.

Reply:
@rafaelcr that is a good suggestion. @zone117x suggested the returning txs be in the order they were requested.
@zone117x what do you think about this approach?

Reply:
Done